cli/lib/png2src.js

```diff
@@ -114,76 +114,82 @@ const ALIASES = {
     cpp: "c",
 };
 
-function run (sourceFile) {
+function run(sourceFile) {
     const png = pngjs.PNG.sync.read(fs.readFileSync(sourceFile), {
         colorType: 1,
         inputColorType: 1,
     });
 
-    if (!png.palette) {
-        throw new Error("Does not have indexed color palette");
-    }
-
-    // Map rgba to palette index
-    const palette = new Map();
-    for (let ii = 0; ii < png.palette.length; ++ii) {
-        const rgba = png.palette[ii];
-        const packed = (rgba[0] << 24) | (rgba[1] << 16) | (rgba[2] << 8) | rgba[3];
-        palette.set(packed, ii);
+    const palette = [];
+    for (let y = 0; y < png.height; ++y) {
+        for (let x = 0; x < png.width; ++x) {
+            const idx = 4 * (png.width * y + x);
+            const r = png.data[idx];
+            const g = png.data[idx + 1];
+            const b = png.data[idx + 2];
+            const a = png.data[idx + 3];
+            const packed = (r << 24) | (g << 16) | (b << 8) | a;
+            let s = palette.findIndex(v => v === packed);
+            if (s === -1) {
+                palette.push(packed);
```

Comment:
Can we abort early here if the palette would grow to larger than 4? We can also print the coordinate of the offending pixel.

Reply:
Sure thing. I'm on it.